File: apps/aberp-ui/ui/src/lib/invoice-dates.ts
// PR-84 — pure date helpers for the issue-invoice form's three-date UX.
//
// Three concepts, each with specific rules per the PR-84 brief:
//
//   1. Invoice date (Számla kelte) — always server-stamped today; never
//      typed by the operator. The form renders it read-only; no helper
//      here computes it (the server's clock is the truth).
//
//   2. Payment deadline (Fizetési határidő) — bidirectional: the
//      operator may type an offset in days (invoice date + N) OR an
//      absolute date, and the two fields update each other live. The
//      helpers `addDays` and `daysBetween` are the round-trip pair.
//
//   3. Delivery / fulfillment date (Teljesítési dátum) — REGULATORY
//      (NAV `invoiceDeliveryDate`; drives the VAT-period assignment).
//      The "comfort zone" is the closed interval [invoice_date,
//      payment_deadline]; out-of-range choices are allowed but pop a
//      soft "Are you sure?" confirmation and are audit-logged.
//
// Pure, no Svelte runes / no DOM — pinned under vitest. Calendar-date
// arithmetic only (YYYY-MM-DD strings); no timestamps, no time zones
// (Hungarian invoice dates are calendar dates, not instants).

/** Wire form for all three invoice dates — canonical ISO `YYYY-MM-DD`.
 * Matches the NAV `xs:date` shape (per `reference_nav_gotchas.md`
 * §"Date format") and the DuckDB `DATE` column form. */
export type IsoDate = string;

const ISO_DATE_RE = /^(\d{4})-(\d{2})-(\d{2})$/;

/** Default offset for the payment-deadline field on a fresh form (per
 * Ervin's HU-business convention: 8 days). The operator can override
 * to any non-negative integer; same-day (offset 0) is permitted and
 * means "due on the invoice date itself" (cash sale). */
export const DEFAULT_PAYMENT_OFFSET_DAYS = 8;

/** Parse an ISO date into a [year, month, day] triple. Returns null on
 * malformed input — the operator-typed input goes through this so the
 * form can surface a precise validation error rather than silently
 * substituting NaN downstream. */
export function parseIsoDate(s: IsoDate): { y: number; m: number; d: number } | null {
  const match = ISO_DATE_RE.exec(s);
  if (!match) return null;
  const y = Number(match[1]);
  const m = Number(match[2]);
  const d = Number(match[3]);
  if (m < 1 || m > 12 || d < 1 || d > 31) return null;
  // UTC noon avoids DST surprises across timezones; we only care about
  // the calendar date. Verify the parsed date round-trips (catches
  // overflow like "2026-02-30" which Date silently corrects).
  const dt = new Date(Date.UTC(y, m - 1, d, 12, 0, 0));
  if (dt.getUTCFullYear() !== y || dt.getUTCMonth() !== m - 1 || dt.getUTCDate() !== d) {
    return null;
  }
  return { y, m, d };
}

/** Format a Date (interpreted in UTC) into a YYYY-MM-DD string. */
function formatIsoFromUtc(dt: Date): IsoDate {
  const y = dt.getUTCFullYear();
  const m = String(dt.getUTCMonth() + 1).padStart(2, "0");
  const d = String(dt.getUTCDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

/** Today's date in the operator's local timezone, rendered as
 * YYYY-MM-DD. Used as the default for the read-only invoice-date
 * field on first paint. The server stamps the true issue date at
 * issuance time; this is a display default only. */
export function todayLocalIso(now: Date = new Date()): IsoDate {
  const y = now.getFullYear();
  const m = String(now.getMonth() + 1).padStart(2, "0");
  const d = String(now.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

/** Add N days to an ISO date and return the resulting ISO date. N may
 * be negative (subtract). Returns null if the input is malformed.
 * Handles month and year boundaries correctly (delegates to JS Date's
 * UTC arithmetic). */
export function addDays(iso: IsoDate, n: number): IsoDate | null {
  const parsed = parseIsoDate(iso);
  if (parsed === null) return null;
  if (!Number.isFinite(n) || !Number.isInteger(n)) return null;
  const dt = new Date(Date.UTC(parsed.y, parsed.m - 1, parsed.d, 12, 0, 0));
  dt.setUTCDate(dt.getUTCDate() + n);
  return formatIsoFromUtc(dt);
}

/** Calendar-day difference `toIso - fromIso`. Positive when `toIso` is
 * after `fromIso`, zero when equal, negative when before. Returns null
 * if either input is malformed. Round-trips with `addDays`:
 * `addDays(from, daysBetween(from, to)) === to`. */
export function daysBetween(fromIso: IsoDate, toIso: IsoDate): number | null {
  const a = parseIsoDate(fromIso);
  const b = parseIsoDate(toIso);
  if (a === null || b === null) return null;
  const da = Date.UTC(a.y, a.m - 1, a.d, 12, 0, 0);
  const db = Date.UTC(b.y, b.m - 1, b.d, 12, 0, 0);
  return Math.round((db - da) / 86_400_000);
}

/** Classification of a candidate delivery date against the
 * [invoice_date, payment_deadline] comfort zone. The form's
 * "Are you sure?" confirm fires for `BeforeInvoiceDate` and
 * `AfterPaymentDeadline`; `InRange` silently accepts the choice. The
 * audit payload stamps the same discriminant so the regulatory trail
 * records every out-of-range override.
 *
 * Bounds are INCLUSIVE — a delivery date equal to either endpoint is
 * in range (no confirm). */
export type ComfortZone = "InRange" | "BeforeInvoiceDate" | "AfterPaymentDeadline";

/** Classify a candidate delivery date against the comfort zone. The
 * comfort zone is the closed interval [invoice_date, payment_deadline].
 *
 * Returns null if any input is malformed OR if `payment_deadline` is
 * before `invoice_date` (a malformed range — the caller's form-level
 * validation should already have caught this, but we refuse to
 * classify against it rather than producing a wrong answer). */
export function comfortZone(
  invoiceDate: IsoDate,
  paymentDeadline: IsoDate,
  deliveryDate: IsoDate,
): ComfortZone | null {
  const beforeInvoice = daysBetween(invoiceDate, deliveryDate);
  const afterDeadline = daysBetween(paymentDeadline, deliveryDate);
  const rangeSpan = daysBetween(invoiceDate, paymentDeadline);
  if (beforeInvoice === null || afterDeadline === null || rangeSpan === null) {
    return null;
  }
  if (rangeSpan < 0) return null;
  if (beforeInvoice < 0) return "BeforeInvoiceDate";
  if (afterDeadline > 0) return "AfterPaymentDeadline";
  return "InRange";
}

/** Wire-form discriminant the audit payload records for the delivery-
 * date choice. `null` means "in range, no override" (default-path,
 * unflagged on the audit row). The two override values mirror the
 * `ComfortZone` enum's out-of-range arms verbatim. */
export type DeliveryDateOverride = "BeforeInvoiceDate" | "AfterPaymentDeadline" | null;

/** Map a `ComfortZone` classification to the audit-wire override
 * discriminant. `InRange` becomes `null`; the two out-of-range arms
 * map to themselves. Centralised so the SPA → backend wire form
 * stays consistent with the audit-payload field. */
export function overrideKindForZone(zone: ComfortZone): DeliveryDateOverride {
  if (zone === "InRange") return null;
  return zone;
}

File: modules/billing/src/domain/invoice_dates.rs
//! PR-84 — pure date-classification helpers for the issue-invoice
//! three-date UX. Mirror of the SPA's `invoice-dates.ts` helpers; lives
//! in the domain layer because the audit stamp records the same
//! classification verbatim (regulatory: out-of-range delivery dates are
//! a tamper-evident operator override).
//!
//! Three concepts (recap):
//!   1. Invoice date — server-stamped today, immutable.
//!   2. Payment deadline — operator picks, fixed or offset.
//!   3. Delivery date — REGULATORY (NAV `invoiceDeliveryDate`); drives
//!      the VAT-period assignment. The "comfort zone" is the closed
//!      interval `[invoice_date, payment_deadline]`; out-of-range
//!      choices flow through but are flagged in the audit ledger.
//!
//! Pure, no IO. Pinned in this module's `#[cfg(test)]` block.

use time::Date;

/// Classification of a candidate delivery date against the comfort
/// zone. The form's "Are you sure?" confirm fires for the two out-of-
/// range arms; the audit payload's `delivery_date_override` field
/// records the same discriminant verbatim (regulatory trail).
///
/// Bounds are INCLUSIVE — a delivery date equal to either endpoint is
/// in range (no confirm, no audit flag).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DeliveryDateZone {
    /// The delivery date falls in `[invoice_date, payment_deadline]`.
    /// The expected operator path; no audit override flag.
    InRange,
    /// The delivery date is before the invoice date. Legitimate
    /// (delivery happened before invoicing) but it shifts the VAT
    /// period earlier than the invoice itself — operator confirms via
    /// the SPA's inline "Are you sure?" affordance and the audit
    /// payload stamps `BeforeInvoiceDate`.
    BeforeInvoiceDate,
    /// The delivery date is after the payment deadline. Legitimate
    /// (advance billing of a future delivery) but it pushes the VAT
    /// period later — operator confirms via the same inline
    /// affordance and the audit payload stamps `AfterPaymentDeadline`.
    AfterPaymentDeadline,
}

impl DeliveryDateZone {
    /// Wire-form discriminant for the audit-ledger payload. The
    /// `InRange` case maps to `None` (default-path; no override flag on
    /// the audit row); the two out-of-range arms map to themselves as
    /// PascalCase strings matching the SPA's `DeliveryDateOverride`
    /// union exactly.
    pub fn audit_discriminant(self) -> Option<&'static str> {
        match self {
            Self::InRange => None,
            Self::BeforeInvoiceDate => Some("BeforeInvoiceDate"),
            Self::AfterPaymentDeadline => Some("AfterPaymentDeadline"),
        }
    }
}

/// Classify a candidate delivery date against the closed-interval
/// comfort zone `[invoice_date, payment_deadline]`.
///
/// Returns `None` when `payment_deadline < invoice_date` — a malformed
/// range the caller's input validator should already have rejected.
/// Refuse to classify rather than producing a wrong answer.
pub fn classify_delivery_date(
    invoice_date: Date,
    payment_deadline: Date,
    delivery_date: Date,
) -> Option<DeliveryDateZone> {
    if payment_deadline < invoice_date {
        return None;
    }
    if delivery_date < invoice_date {
        return Some(DeliveryDateZone::BeforeInvoiceDate);
    }
    if delivery_date > payment_deadline {
        return Some(DeliveryDateZone::AfterPaymentDeadline);
    }
    Some(DeliveryDateZone::InRange)
}

#[cfg(test)]
mod tests {
    //! Heavy boundary coverage — the comfort-zone classifier is the
    //! load-bearing seam between operator UX and the regulatory audit
    //! trail. A regression that flips an inclusive bound to exclusive
    //! would silently turn an in-range choice into an audit-flagged
    //! override (or vice versa); pin both endpoints + the single-day
    //! edge case explicitly.

    use super::*;
    use time::macros::date;

    #[test]
    fn in_range_when_delivery_equals_invoice_date() {
        // Left endpoint inclusive.
        let zone = classify_delivery_date(
            date!(2026 - 05 - 27),
            date!(2026 - 06 - 04),
            date!(2026 - 05 - 27),
        )
        .unwrap();
        assert_eq!(zone, DeliveryDateZone::InRange);
        assert_eq!(zone.audit_discriminant(), None);
    }

    #[test]
    fn in_range_when_delivery_equals_payment_deadline() {
        // Right endpoint inclusive.
        let zone = classify_delivery_date(
            date!(2026 - 05 - 27),
            date!(2026 - 06 - 04),
            date!(2026 - 06 - 04),
        )
        .unwrap();
        assert_eq!(zone, DeliveryDateZone::InRange);
        assert_eq!(zone.audit_discriminant(), None);
    }

    #[test]
    fn in_range_when_delivery_is_strictly_inside() {
        let zone = classify_delivery_date(
            date!(2026 - 05 - 27),
            date!(2026 - 06 - 04),
            date!(2026 - 05 - 30),
        )
        .unwrap();
        assert_eq!(zone, DeliveryDateZone::InRange);
    }

    #[test]
    fn before_invoice_date_when_delivery_is_one_day_earlier() {
        let zone = classify_delivery_date(
            date!(2026 - 05 - 27),
            date!(2026 - 06 - 04),
            date!(2026 - 05 - 26),
        )
        .unwrap();
        assert_eq!(zone, DeliveryDateZone::BeforeInvoiceDate);
        assert_eq!(zone.audit_discriminant(), Some("BeforeInvoiceDate"));
    }

    #[test]
    fn before_invoice_date_when_delivery_is_months_earlier() {
        let zone = classify_delivery_date(
            date!(2026 - 05 - 27),
            date!(2026 - 06 - 04),
            date!(2026 - 01 - 10),
        )
        .unwrap();
        assert_eq!(zone, DeliveryDateZone::BeforeInvoiceDate);
    }

    #[test]
    fn after_payment_deadline_when_delivery_is_one_day_later() {
        let zone = classify_delivery_date(
            date!(2026 - 05 - 27),
            date!(2026 - 06 - 04),
            date!(2026 - 06 - 05),
        )
        .unwrap();
        assert_eq!(zone, DeliveryDateZone::AfterPaymentDeadline);
        assert_eq!(zone.audit_discriminant(), Some("AfterPaymentDeadline"));
    }

    #[test]
    fn after_payment_deadline_when_delivery_is_far_in_the_future() {
        let zone = classify_delivery_date(
            date!(2026 - 05 - 27),
            date!(2026 - 06 - 04),
            date!(2027 - 01 - 01),
        )
        .unwrap();
        assert_eq!(zone, DeliveryDateZone::AfterPaymentDeadline);
    }

    #[test]
    fn zero_day_comfort_zone_is_single_day_inclusive() {
        // Cash sale: invoice issued + due same day. The point-interval
        // [d, d] is still inclusive, so delivery == that day is in range
        // and the day before / after both trigger the override.
        let same = date!(2026 - 05 - 27);
        assert_eq!(
            classify_delivery_date(same, same, same).unwrap(),
            DeliveryDateZone::InRange
        );
        assert_eq!(
            classify_delivery_date(same, same, date!(2026 - 05 - 26)).unwrap(),
            DeliveryDateZone::BeforeInvoiceDate
        );
        assert_eq!(
            classify_delivery_date(same, same, date!(2026 - 05 - 28)).unwrap(),
            DeliveryDateZone::AfterPaymentDeadline
        );
    }

    #[test]
    fn malformed_range_returns_none() {
        // payment_deadline < invoice_date — the form-level validator
        // should already have caught this; the classifier refuses to
        // produce a misleading answer rather than papering over the bug.
        assert_eq!(
            classify_delivery_date(
                date!(2026 - 06 - 04),
                date!(2026 - 05 - 27),
                date!(2026 - 05 - 30),
            ),
            None,
        );
    }
}
